### PR TITLE
(PDB-1534) Split command and version into their own services

### PR DESCRIPTION
**Pull request by ajroetker**
This commit splits out the command and version endpoints into their own
trapperkeeper services with their own versioned APIs. The version
endpoint is hosted at `/pdb/version/v1` and the command endpoint is
served at `/pdb/cmd/v1`.

### CODE REVIEW

**Comment by ajroetker** [comment]
@pljenkinsro retest this please

**Comment by mullr** [comment]
Acceptance tests are still expecting the version service at  http://localhost:8080/pdb/query/v4/version

**Comment by ajroetker** [comment]
@pljenkinsro retest this please

**Comment by ajroetker** [comment]
@pljenkinsro retest this please

**Comment by wkalt** [comment]
@pljenkinsro retest this please

File: documentation/_puppetdb_nav.html [hunk 1]
```diff
@@ -78,7 +83,11 @@ <h2 id="puppetdb">PuppetDB NEXT</h2>
       <li><a href="/puppetdb/{{ puppetdb_version }}/api/query/v4/event-counts.html">Event Counts Endpoint</a></li>
       <li><a href="/puppetdb/{{ puppetdb_version }}/api/query/v4/aggregate-event-counts.html">Aggregate Event Counts Endpoint</a></li>
       <li><a href="/puppetdb/{{ puppetdb_version }}/api/query/v4/server-time.html">Server Time Endpoint</a></li>
-      <li><a href="/puppetdb/{{ puppetdb_version }}/api/query/v4/version.html">Version Endpoint</a></li>
+    </ul>
+  </li>
+  <li><strong>Version API Version 1</strong>
+    <ul>
+      <li><a href="/puppetdb/{{ puppetdb_version }}/api/version/v1/verison.html">Command API</a></li>
```
Comment (wkalt) [inline documentation/_puppetdb_nav.html]: verison

Comment (wkalt) [inline documentation/_puppetdb_nav.html]: also these links need to change to meta now right?

File: src/puppetlabs/puppetdb/cli/services.clj [hunk 2]
```diff
@@ -193,26 +191,33 @@
...
     (log/info "Shutdown request received; puppetdb exiting."))
-  (when-let [updater (context :updater)]
-    (log/info "Shutting down updater thread.")
-    (future-cancel updater))
   (shutdown-mq context)
-  (when-let [ds (get-in context [:shared-globals :scf-write-db :datasource])]
-    (.close ds))
-  (when-let [ds (get-in context [:shared-globals :scf-read-db :datasource])]
-    (.close ds))
+  (shutdown-updater context)
```
Comment (wkalt) [inline src/puppetlabs/puppetdb/cli/services.clj]: you've changed the shutdown ordering here -- I don't think it matters but was there a reason for that?

File: src/puppetlabs/puppetdb/cli/services.clj [hunk 3]
```diff
@@ -221,16 +226,16 @@
...
          (map? config)
          (ifn? add-ring-handler)
          (ifn? shutdown-on-error)]
    :post [(map? %)
-          (every? (partial contains? %) [:broker])]}
+          (contains? % :broker)]}
   (let [{:keys [global   jetty
                 database read-database
-                puppetdb command-processing]} (conf/process-config! config)
+                puppetdb command-processing]} config
```
Comment (wkalt) [inline src/puppetlabs/puppetdb/cli/services.clj]: trailing whitespace

File: src/puppetlabs/puppetdb/cli/services.clj [hunk 4]
```diff
@@ -239,17 +244,12 @@
...
-        mq-connection-str (add-max-framesize max-frame-size mq-addr)
-        authorizer (when certificate-whitelist
-                     (build-whitelist-authorizer certificate-whitelist))]
+        read-db (pl-jdbc/pooled-datasource read-database)
+        puppet-enterprise? (not= product-name "puppetdb")]
 
-    (when-let [v (version)]
-      (log/infof "PuppetDB version %s" v))
+    (some->> (v/version)
+             (log/infof "PuppetDB version %s"))
```
Comment (wkalt) [inline src/puppetlabs/puppetdb/cli/services.clj]: I don't think this change buys us anything or relates to this PR.

File: src/puppetlabs/puppetdb/cli/services.clj [hunk 5]
```diff
@@ -259,83 +259,81 @@
...
+          ;; Run database maintenance tasks seqentially to avoid
+          ;; competition. Each task must handle its own errors.
+          (gc-task db-maintenance-tasks)
+          (gc-task (fn [] (compress-dlo! dlo-compression-threshold discard-dir))))
+        (assoc context
+               :broker broker
+               :mq-factory mq-factory
+               :mq-connection mq-connection
+               :shared-globals globals
+               :updater updater)))))
```
Comment (wkalt) [inline src/puppetlabs/puppetdb/cli/services.clj]: I'm having a hard time understanding what has changed in this function. Is it just a refactor?

File: src/puppetlabs/puppetdb/cli/services.clj [hunk 6]
```diff
@@ -347,19 +345,24 @@
...
-        (stop-puppetdb context false))
-  (shared-globals [this]
-                  (:shared-globals (service-context this)))
+         (let [config (-> (get-config)
+                          conf/process-config!)
+               url-prefix (get-route this)
+               shutdown-on-error* (partial shutdown-on-error (service-id this))
+               add-ring-handler* #(->> %
+                                       (compojure/context url-prefix [])
+                                       (add-ring-handler this))]
```
Comment (wkalt) [inline src/puppetlabs/puppetdb/cli/services.clj]: similar question here -- is this just moving function calls out of start-puppetdb and into puppetdb-service or is there something else going on?

File: src/puppetlabs/puppetdb/cli/services.clj (hunk 5 above)
Comment (ajroetker) [inline src/puppetlabs/puppetdb/cli/services.clj]: So I was trying to refactor things to make it easier to split out setting up the mq connection from the rest of config processing. I was also trying to refactor this to see what all was going on when we try to shut down puppetdb, to see what all should be moved to the shut-down of the command service. But yes for the most part just a refactor.

File: src/puppetlabs/puppetdb/cli/services.clj (hunk 6 above)
Comment (ajroetker) [inline src/puppetlabs/puppetdb/cli/services.clj]: The idea here was separating out anything that needs `this` from the start-puppetdb function to make it a little less monolithic and see what kinds of trapperkeeper related work we're doing from the defservice.

**Comment by ajroetker** [comment]
@pljenkinsro retest this please

File: src/puppetlabs/puppetdb/cli/services.clj (hunk 2 above)
Comment (ajroetker) [inline src/puppetlabs/puppetdb/cli/services.clj]: no reason, just an accident I'll move it back

File: src/puppetlabs/puppetdb/http/event_counts.clj [hunk 7]
```diff
@@ -9,27 +9,23 @@
...
-              [summarize_by query-options paging-options]
-              (:scf-read-db globals)
-              (:url-prefix globals))))}))
+  [globals]
+  (let [{:keys [api-version url-prefix scf-read-db]} globals]
+   (app
+    [""]
+    {:get (fn [{:keys [params paging-options]}]
+            (let [{:strs [query summarize_by counts_filter count_by] :as query-params} params
+                  query-options (merge {:counts_filter (some-> counts_filter (json/parse-strict-string true))
```
Comment (wkalt) [inline src/puppetlabs/puppetdb/http/event_counts.clj]: I think when is more readable than some-> in these cases

Comment (ajroetker) [inline src/puppetlabs/puppetdb/http/event_counts.clj]: Aren't they the same thing in these cases? Is it just the case that `some` is unfamiliar?

**Comment by wkalt** [comment]
@ajroetker I'd really prefer if the refactoring and style edits were left out of this PR to the greatest extent possible (understanding that sometimes it's unavoidable / clearly the right choice.) The PR is pretty massive even without them, they make it harder to detect functional changes, and I disagree with a number of them even though I agree that we need the functionality that is the objective of the PR. Might be too tangled up to easily change at this point, but that's my general stance on these things.

**Comment by ajroetker** [comment]
As far as the refactors in services.clj @wkalt, I had originally split out the command service and started to pick apart the config processing/shutdown so that some of that could be moved to the PuppetDBCommandService itself, after making the changes the try and move the management of the mq to the CommandService, the PR was starting to get quite lofty and so I left it as it was so that the actual substantive changes to the service management could be more modular. Any other refactors done in this PR were purely for the sake of keeping our various trapperkeeper services consistent with one another.

**Comment by wkalt** [comment]
@ajroetker I trust your best judgement -- just wanted to make that point

**Comment by wkalt** [comment]
@ajroetker looks like the bernchmark tool is trying to hit /pdb/query/v4?checksum=blah for command submission

**Comment by wkalt** [comment]
we also need to update the upgrading-from-v3 doc to describe the pathing changes

**Comment by wkalt** [comment]
also, you'll see a couple spots in the docs to fix if you grep the codebase for "v4\/commands"

**Comment by wkalt** [comment]
likewise for "v4\/version"

**Comment by wkalt** [comment]
one of those commands instances in in basic_fact_retrieval.rb -- I'd expect that to cause the acceptance tests to fail

**Comment by wkalt** [comment]
server-time.markdown needs to be brought up to date

**Comment by wkalt** [comment]
+1 aside from outstanding stuff

**Comment by ajroetker** [comment]
@pljenkinsro retest this please

**Comment by ajroetker** [comment]
@pljenkinsro retest this please

**Comment by ajroetker** [comment]
waiting to rebase after @mullr comments

**Comment by ajroetker** [comment]
@pljenkinsro retest this please

**Comment by mullr** [comment]
:-1: here. I've made it about 1/3 of the way through, and it's just too much. What I've been doing to untangle the moves vs. changes is keeping two diffs open, trying to find the old location and the new location of something in each, and flipping back and forth so I can see what changed. Even this barely works, as the change is large enough that github's code review JS starts to fall over. 

This needs to be split into separate commits. This will seem like a lot of work, but actually it's going to be about the same amount of work as I'm doing trying to untangle the structure in my head - the difference is that you already know what the structure is. It doesn't need to be super detailed - the main problem is the splitting of command.clj and its associated tests. There should be one commit which does the splitting, updating references as appropriate, and another which changes the functionality. It looks like there are some other big-lump changes in the tests which could use the same treatment.

**Comment by ajroetker** [comment]
I'm just going to start from scratch here then, I'll make different tickets for the pieces involved since this probably was too much work for one ticket in the first place. Also if you could comment on the PR next time it would be nice because then I don't have to worry about overriding your comments if i force push.

**Comment by mullr** [comment]
yeah that was a misclick... for like the 50th time. I can help you split
this up if you're worried about git-fu or anything.

**Comment by ajroetker** [comment]
I'm not worried about git-fu but I'm worried that it will take longer than simply rewriting it, also the way it is I'll be having to rebase on almost any patch that comes in and I'd have to structure everything as to make those rebases easier.

**Comment by mullr** [comment]
this is a 3k line patch... how can rewriting it be faster? anway, it's up
to you.

**Comment by ajroetker** [comment]
I'm gonna rewrite the first part `/pdb/cmd` and leave the rest as open tickets

**Pull request closed without merging**